File: app/src/main/java/com/bogdan/codeforceswatcher/features/users/FirebaseController.kt
package com.bogdan.codeforceswatcher.features.users

import com.google.firebase.auth.FirebaseAuth
import io.xorum.codeforceswatcher.features.auth.IFirebaseController
import kotlin.Exception

class FirebaseController : IFirebaseController {
    private val auth = FirebaseAuth.getInstance()

    override fun signIn(email: String, password: String, callback: (Exception?) -> Unit) {
        auth.signInWithEmailAndPassword(email, password).addOnSuccessListener {
            callback(null)
        }.addOnFailureListener { e ->
            callback(e)
        }
    }

    override fun signUp(email: String, password: String, callback: (Exception?) -> Unit) {
        auth.createUserWithEmailAndPassword(email, password).addOnSuccessListener {
            callback(null)
        }.addOnFailureListener { e ->
            callback(e)
        }
    }

    override fun fetchToken(callback: (String?, Exception?) -> Unit) {
        val firebaseUser = auth.currentUser ?: run {
            callback(null, null)
            return@fetchToken
        }
        firebaseUser.getIdToken(false).addOnSuccessListener { task ->
            callback(task.token, null)
        }.addOnFailureListener { e ->
            callback(null, e)
        }
    }

    override fun logOut(callback: (Exception?) -> Unit) {
        auth.signOut()
        callback(null)
    }

    override fun sendPasswordReset(email: String, callback: (Exception?) -> Unit) {
        auth.sendPasswordResetEmail(email).addOnSuccessListener {
            callback(null)
        }.addOnFailureListener { e ->
            callback(e)
        }
    }
}

File: ios/Codeforces Watcher/Application/utils/FirebaseController.swift
//
//  FirebaseController.swift
//  Codeforces Watcher
//
//  Created by Den Matiash on 17.03.2021.
//  Copyright © 2021 xorum.io. All rights reserved.
//

import Foundation
import FirebaseAuth
import common

class FirebaseController: IFirebaseController {
    
    private var auth: Auth { Auth.auth() }
    
    func signIn(email: String, password: String, callback: @escaping (String?, KotlinException?) -> Void) {
        auth.signIn(withEmail: email, password: password) { authResult, e in
            if let e = e {
                callback(nil, e.toKotlinException())
                return
            }
            self.fetchToken { token, e in
                if let token = token {
                    callback(token, nil)
                } else {
                    callback(nil, e)
                }
            }
        }
    }
    
    func signUp(email: String, password: String, callback: @escaping (String?, KotlinException?) -> Void) {
        auth.createUser(withEmail: email, password: password) { authResult, e in
            if let e = e {
                callback(nil, e.toKotlinException())
                return
            }
            self.fetchToken { token, e in
                if let token = token {
                    callback(token, nil)
                } else {
                    callback(nil, e)
                }
            }
        }
    }
    
    func fetchToken(callback: @escaping (String?, KotlinException?) -> Void) {
        guard let _ = auth.currentUser else {
            callback(nil, nil)
            return
        }
        
        auth.currentUser?.getIDToken { token, e in
            if let token = token {
                callback(token, nil)
            } else {
                callback(nil, e?.toKotlinException())
            }
        }
    }
    
    func logOut(callback: @escaping (KotlinException?) -> Void) {
        do {
            try auth.signOut()
            callback(nil)
        } catch let e as NSError {
            callback(e.toKotlinException())
        }
    }
    
    func sendPasswordReset(email: String, callback: @escaping (KotlinException?) -> Void) {
        auth.sendPasswordReset(withEmail: email) { e in
            if let e = e {
                callback(e.toKotlinException())
            } else {
                callback(nil)
            }
        }
    }
}

fileprivate extension Error {
    
    func toKotlinException() -> KotlinException {
        KotlinException(message: self.localizedDescription)
    }
}
